Kind ansible CI

Mmduh-483 opened the pull request:
This PR presents a new CI structure using kind and ansible

- Present new script run_kind_ci.sh to run CI using kind cluster 
- Divide CI into 7 phases ("prepare-ci-environment" "deploy-kind" "utilities" "deploy-project" "test" "undeploy-project" "undeploy-kind")
- Run each CI with its corresponding ansible play-book

Mmduh-483 committed:
Install ansible in prepare_setup.sh

Signed-off-by: Mamduh Alassi <mamduhala@nvidia.com>

Mmduh-483 committed:
Add install install/update kind latest upstream release with ansible

Signed-off-by: Mamduh Alassi <mamduhala@nvidia.com>

Mmduh-483 committed:
Add deploy kind cluster

--num-workers number of cluster workers
--kind-config kind config file to use, when used ignores num-workers
--kubeconfig kubeconfig for cluster

Signed-off-by: Mamduh Alassi <mamduhala@nvidia.com>

Mmduh-483 committed:
Add remove kind cluster phase

Signed-off-by: Mamduh Alassi <mamduhala@nvidia.com>

abdallahyas commented on ansible/roles/kind/kind_cluster_create/tasks/main.yml:
```diff
@@ -0,0 +1,12 @@
+  block:
+    - template:
+        src: kind.yaml.j2
+        dest: /tmp/kind.yaml
```
I think it would be good if this is added to the `artifacts` directory instead of the tmp, so that it will be added to the logs.

ykulazhenkov replied:
I agree, It will be great to have kind config as a build artifact. 
Also, there is a `tempfile` module for ansible [1] which can be used for tmp files generation.
[1] https://docs.ansible.com/ansible/2.9/modules/tempfile_module.html

Mmduh-483 replied:
Added

abdallahyas commented on ansible/roles/kind/kind_cluster_create/tests/inventory:
```diff
@@ -0,0 +1,2 @@
+localhost
```
maybe use the `kind_nodes` group? so that it will be consistent and we will not have any surprises in the future in case kind_nodes is changed.

Mmduh-483 replied:
In playbook `kind_nodes` group is used, the `role/tests` is removed now since it is not needed

abdallahyas commented on ansible/deploy-kind.yaml:
```diff
@@ -0,0 +1,8 @@
+- name: Deploy Kind Cluster
+  hosts: kind_nodes
+  become: true
+  remote_user: root
```
all our CIs runs in stack user and use sudo to perform privileged tasks, so this have a very big chance that it will fail when run using a Jenkins job.

I think it would be best to use the [become](https://docs.ansible.com/ansible/latest/user_guide/become.html) module and only for it to be used when root privilege is needed.

Mmduh-483 replied:
`become` is already used in each phase play-book, I tested it with stack user on CI machine and it is working fine

abdallahyas commented on ansible/roles/kind/kind_cluster_create/tasks/main.yml:
```diff
@@ -0,0 +1,12 @@
+  when: 'kind_config == ""'
+- name: Deploy Kind cluster
+  shell:
+    cmd: kind create cluster --name {{project}} --kubeconfig {{kubeconfig}} --config {{kind_config}}
```
What will happen if a config file was passed to the playbook, and the config file contains the name field, what name will take precedence?

I think since the config file is always used (whether it was passed or rendered) it would be a good idea to add the `name` field to the jinja file so that all cluster configs are in one place (except the kubeconfig since it looks like it is not configurable using the config file).

Mmduh-483 replied:
`--name` overrides the name in config file, it is intended to name the cluster with the ci project name

abdallahyas commented on ansible/roles/kind/kind_cluster_create/tests/test.yml:
```diff
@@ -0,0 +1,5 @@
+---
+- hosts: localhost
+  remote_user: root
```
same as above regarding the `become`

abdallahyas commented on ansible/roles/kind/kind_cluster_create/vars/main.yml:
```diff
@@ -0,0 +1,5 @@
+---
+project: "{{ lookup('env', 'PROJECT') }}"
```
since the `project` variable is used across playbooks to access the cluster name, maybe use group vars instead of defining it per role?

Mmduh-483 replied:
added

abdallahyas commented on ansible/roles/kind/kind_cluster_destroy/tests/inventory:
```diff
@@ -0,0 +1,2 @@
+localhost
```
same as above regarding the using the `kind_node` group

abdallahyas commented on ansible/roles/kind/kind_cluster_destroy/tests/test.yml:
```diff
@@ -0,0 +1,5 @@
+---
+- hosts: localhost
+  remote_user: root
```
same as above regarding the `become`

abdallahyas commented on ansible/roles/kind/kind_install/tests/test.yml:
```diff
@@ -0,0 +1,5 @@
+---
+- hosts: localhost
+  remote_user: root
```
same as above regarding the `become`

abdallahyas commented on ansible/roles/kind/kind_install/tests/inventory:
```diff
@@ -0,0 +1,2 @@
+localhost
```
same as above regarding the group

abdallahyas commented on run_kind_ci.sh:
```diff
@@ -0,0 +1,137 @@
+#!/bin/bash
+
+KIND_CI_PROJECTS=("sriov-cni" "sriov-ib" "antrea" "ipoib" "network-operator" "antrea")
```
antrea is duplicated, can you remove the duplication?

Mmduh-483 replied:
fixed

abdallahyas commented on run_kind_ci.sh:
```diff
@@ -0,0 +1,137 @@
+    --skip-phases)
+      shift
+      SKIP_PHASES=$1
+      IFS="," read -a SKIP_PHASES <<<"$1"
+      for phase in "${SKIP_PHASES[@]}"; do
+        containsElement "$phase" "${KIND_CI_PHASES[*]}"
+        if [[ "$?" != "0" ]]; then
+          echo "Unknown phase $phase"
+          usage
+          exit 1
+        fi
+      done
+
+      PHASES_TO_RUN=()
+      for phase in "${SUPPORTED_PHASES[@]}"; do
+        containsElement "$phase" "${SKIP_PHASES[*]}"
+        if [[ "$?" == "1" ]]; then
+          PHASES_TO_RUN+=("$phase")
+        fi
+      done
```
i think it would be a good idea to have this in a function. maybe define the functions before reading inputs.

Mmduh-483 replied:
added a function

ykulazhenkov commented on ansible/inventory/hosts:
```diff
@@ -0,0 +1,2 @@
+[kind_nodes]
+localhost ansible_connection=local
```
Is there any reason why there are no newlines at the end of some files? If no, could you please add them?

Mmduh-483 replied:
added new line to files

ykulazhenkov commented on ansible/roles/kind/kind_cluster_create/meta/main.yml:
```diff
@@ -0,0 +1,53 @@
+galaxy_info:
```
Do you plan to fill galaxy_info in metafiles for all roles? I think these files are not required when roles used locally and can be safely dropped if we don't plan to upload roles to the ansible galaxy.

Mmduh-483 replied:
no, I removed it

ykulazhenkov commented on prepare_setup.sh:
```diff
@@ -270,6 +271,18 @@ packages_install(){
+    echo "Installing pip packages $packages ...."
+
+    pip3 install -U pip setuptools
+    pip3 install $packages
```
Is it expected always to use the latest ansible version, or should we use a fixed version here?

Mmduh-483 replied:
I will pin it to the current version 3.2.0

ykulazhenkov commented on prepare_setup.sh:
```diff
@@ -299,6 +312,9 @@ main(){
     packages_install
     let status=$status+$?
 
+    pip_packages_install
```
Do we need to continue installation if `package_install` step failed?

Mmduh-483 replied:
The script behaviour is to continue when even if there is a failure, it would be good to refactor it in another PR

Mmduh-483 committed:
Add prepare setup phase for preparing environment for CI

- Add script run_kind_ci.sh script to run kind based CIs
- Add play book to create workspace and artifacts directories
- Add --skip-phases to skip specific phases
- Add required --project to choose CI project

Signed-off-by: Mamduh Alassi <mamduhala@nvidia.com>

abdallahyas reviewed (approved):
/lgtm

Mmduh-483 commented:
@AbdYsn  @ykulazhenkov can pls merge the PR
